src/main/java/appeng/fluids/parts/FluidHandlerAdapter.java

```diff
@@ -0,0 +1,301 @@
+ *
+ * Applied Energistics 2 is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Applied Energistics 2.  If not, see <http://www.gnu.org/licenses/lgpl>.
+ */
+
+package appeng.fluids.parts;
+
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+import net.minecraftforge.fluids.FluidStack;
+import net.minecraftforge.fluids.capability.IFluidHandler;
+import net.minecraftforge.fluids.capability.IFluidTankProperties;
+
+import appeng.api.AEApi;
+import appeng.api.config.Actionable;
+import appeng.api.networking.security.IActionSource;
+import appeng.api.networking.storage.IBaseMonitor;
+import appeng.api.networking.ticking.TickRateModulation;
+import appeng.api.storage.IMEInventory;
+import appeng.api.storage.IMEMonitorHandlerReceiver;
+import appeng.api.storage.IStorageChannel;
+import appeng.api.storage.channels.IFluidStorageChannel;
+import appeng.api.storage.data.IAEFluidStack;
+import appeng.api.storage.data.IItemList;
+import appeng.me.GridAccessException;
+import appeng.me.helpers.IGridProxyable;
+import appeng.me.storage.ITickingMonitor;
+import appeng.util.item.AEFluidStack;
+
+
+/**
+ * Wraps an Fluid Handler in such a way that it can be used as an IMEInventory for fluids.
+ *
+ * @author BrockWS
+ * @version rv6 - 22/05/2018
+ * @since rv6 22/05/2018
+ */
+public class FluidHandlerAdapter implements IMEInventory<IAEFluidStack>, IBaseMonitor<IAEFluidStack>, ITickingMonitor
+{
+	private final Map<IMEMonitorHandlerReceiver<IAEFluidStack>, Object> listeners = new HashMap<>();
+	private IActionSource source;
+	private final IFluidHandler fluidHandler;
+	private final IGridProxyable proxyable;
+	private final FluidHandlerAdapter.InventoryCache cache;
+
+	FluidHandlerAdapter( IFluidHandler fluidHandler, IGridProxyable proxy )
+	{
+		this.fluidHandler = fluidHandler;
+		this.proxyable = proxy;
+		this.cache = new FluidHandlerAdapter.InventoryCache( this.fluidHandler );
+	}
+
+	@Override
+	public IAEFluidStack injectItems( IAEFluidStack input, Actionable type, IActionSource src )
+	{
+		FluidStack fluidStack = input.getFluidStack();
+
+		// Insert
+		// TODO: May need to brute force like ItemHandlerAdapter
+		int wasFillled = this.fluidHandler.fill( fluidStack, type != Actionable.SIMULATE );
+		int remaining = fluidStack.amount - wasFillled;
+		if( fluidStack.amount == remaining )
+		{
+			// The stack was unmodified, target tank is full
+			return input;
+		}
+
+		if( type == Actionable.MODULATE )
+		{
+			try
+			{
+				this.proxyable.getProxy().getTick().alertDevice( this.proxyable.getProxy().getNode() );
+			}
+			catch( GridAccessException ex )
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Either log or rethrow this exception. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3AS1166)